design-documents/web-api-controllers.md

```diff
@@ -0,0 +1,374 @@
+arguments and results of an operation.
+ 
+Example on how would we change DTOs for operations regarding customer creation/update:
+ 
+First we would have DTO that describes a customer in the way it's convenient for client code when updating a customer:
+```php
+interface CustomerUpdateInterface
+{
+    //No getId() since storefront clients must not be able to control their customer's ID
+    ...
+    
+    //For setting new password
+    public function getPassword(): ?string;
+    
+    public function getAddresses(): CustomerAddress[];
+}
+```
+ 
+Then we would have DTO for customer's addresses:
+```php
+interface CustomerAddressUpdateInterface
+{
+    public function getId(): ?string;
+    ....
+    
+    //No getCustomerId() because we don't give control over that
+    ....
+    
+    public function getPostcode(): string
+}
+```
+
+Since we don't want original password in customer responses we would not return it as read operations result:
+```php
+interface CustomerReadInterface
+{
+    //Now we have getId() since it's OK for storefront clients to know their customer's ID.
+    public function getId(): string;
+    
+    ....
+    //No getPassword()
+    ....
+}
+```
+ 
+By having these operation-specific DTOs we would only have the data we need as arguments and responses.
+ 
+ 
+#### Use declarative validation
+Web API clients displaying forms should be able to get a list of invalid properties sent to display to end users or
+even have validation rules provided to them. Right now entities in Magento are being validated manually and errors
+are being displayed one at the time. Having validation rules provided with the endpoints configuration would solve these
+issues.
+ 
+We could reuse existing validation mechanism that is partially being used for validating customers provided in
+Magento framework in _Magento\Framework\Validator_ namespace
+_(see app/code/Magento/Customer/etc/validation.xml)_.
+ 
+_webapi.xml_ schema could be updated to allow following configuration:
+```xml
+<route url="/V1/customers" method="POST">
+    <service class="Magento\Customer\WebAPI\Controller\CustomerControllerInterface" method="create"/>
+    <resources>
+        <resource ref="anonymous"/>
+    </resources>
+    <validation>
+        <parameters>
+            <parameter name="customer" entity="customer" group="web_api_create">
+                <properties>
+                    <!-- Individual address validators -->
+                    <property name="addresses" entity="customer_address" group="web_api_create_address" walk="true" />
+                    <!-- Validators for the whole collection -->
+                    <property name="addresses" entity="customer_addresses" group="web_api_create" walk="false" />
+                </properties>
+            </parameter>
+        </parameters>
+    </validation>
+</route>
+```
+  
+REST/SOAP framework then would create validator for entity _"customer"_ and group _"web_api_create"_, perform validation
```

Comment:
How web API framework would create validator? Will it be generic framework logic based on declaration above or each validator will have to be implemented separately and just referenced from the declaration?

Reply:
it would use Magento\Framework\Validator\Factory to create validators based on the configuration (groups, entity, constaints defined)